crates/wasi-common/js-polyfill/assets/wasi.js

```diff
@@ -0,0 +1,758 @@
+  Module["GUEST_HEAPU32"] = GUEST_HEAPU32 = new Uint32Array(GUEST_buffer);
+  Module["GUEST_HEAPF32"] = GUEST_HEAPF32 = new Float32Array(GUEST_buffer);
+  Module["GUEST_HEAPF64"] = GUEST_HEAPF64 = new Float64Array(GUEST_buffer);
+}
+
+function copyin_bytes(src, len) {
+  let dst = _malloc(len);
+  updateGuestBuffer();
+
+  for (let i = 0; i < len; ++i) {
+    HEAP8[dst + i] = GUEST_HEAP8[src + i];
+  }
+  return dst;
+}
+
+function copyout_bytes(dst, src, len) {
+  updateGuestBuffer();
+
+  for (let i = 0; i < len; ++i) {
+    GUEST_HEAP8[dst + i] = HEAP8[src + i];
+  }
+  _free(src);
+}
+
+function copyout_i32(dst, src) {
+  updateGuestBuffer();
+
+  GUEST_HEAP32[dst >> 2] = HEAP32[src >> 2];
+  _free(src);
+}
+
+function copyout_i64(dst, src) {
+  updateGuestBuffer();
+
+  GUEST_HEAP32[dst >> 2] = HEAP32[src >> 2];
+  GUEST_HEAP32[(dst + 4) >> 2] = HEAP32[(src + 4) >> 2];
+  _free(src);
+}
+
+function translate_ciovs(iovs, iovs_len) {
+  host_iovs = _malloc(8 * iovs_len);
+  updateGuestBuffer();
+
+  for (let i = 0; i < iovs_len; ++i) {
+    let ptr = GUEST_HEAP32[(iovs + i * 8 + 0) >> 2];
+    let len = GUEST_HEAP32[(iovs + i * 8 + 4) >> 2];
+    let buf = copyin_bytes(ptr, len);
+    HEAP32[(host_iovs + i * 8 + 0) >> 2] = buf;
+    HEAP32[(host_iovs + i * 8 + 4) >> 2] = len;
+  }
+  return host_iovs;
+}
+
+function free_ciovs(host_iovs, iovs_len) {
+  for (let i = 0; i < iovs_len; ++i) {
+    let buf = HEAP32[(host_iovs + i * 8 + 0) >> 2];
+    _free(buf);
+  }
+  _free(host_iovs);
+}
+
+function translate_iovs(iovs, iovs_len) {
+  host_iovs = _malloc(8 * iovs_len);
+  updateGuestBuffer();
+
+  for (let i = 0; i < iovs_len; ++i) {
+    let len = GUEST_HEAP32[(iovs + i * 8 + 4) >> 2];
+    let buf = _malloc(len);
+    updateGuestBuffer();
+    HEAP32[(host_iovs + i * 8 + 0) >> 2] = buf;
+    HEAP32[(host_iovs + i * 8 + 4) >> 2] = len;
+  }
+  return host_iovs;
+}
+
+function free_iovs(host_iovs, iovs_len, iovs) {
+  updateGuestBuffer();
+  for (let i = 0; i < iovs_len; ++i) {
+    let buf = HEAP32[(host_iovs + i * 8 + 0) >> 2];
+    let len = HEAP32[(host_iovs + i * 8 + 4) >> 2];
+    let ptr = GUEST_HEAP32[(iovs + i * 8 + 0) >> 2];
+    copyout_bytes(ptr, buf, len);
+  }
+  _free(host_iovs);
+}
+
+var WASIPolyfill = {
+  args_get: function(argv, argv_buf) {
+    return 0;
+  },
+
+  args_sizes_get: function(argc, argv_buf_size) {
+    updateGuestBuffer();
+
+    // TODO: Implement command-line arguments.
+    GUEST_HEAP32[argc >> 2] = 0;
+    GUEST_HEAP32[argv_buf_size >> 2] = 0;
+    return 0;
+  },
+
+  clock_res_get: function(clock_id, resolution) {
+    let host_resolution = _malloc(8);
+    let ret = _wasi_common_clock_res_get(
+      _get_wasi_context(),
+      HEAP8,
+      HEAP8.length,
+      clock_id,
+      host_resolution
+    );
+    copyout_i64(resolution, host_resolution);
+    return ret;
+  },
+
+  clock_time_get: function(clock_id, precision, time) {
```

Comment:
`precision` is a 64-bit integer, how are you handling that in this project?

Reply:
So far, we're just not handling it. The C version of this polyfill didn't handle it either. We should fix that, but that can be separate.